background/score/workflows_test.go

```diff
@@ -0,0 +1,385 @@
+		})
+
+	ts.env.OnActivity(ts.worker.RefreshLocationStateActivity, mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(
+		func(ctx context.Context, accountNumber, poiID string, metric schema.Metric) (*NotificationProfile, error) {
+			ts.Equal(ts.testAccountNumber, accountNumber)
+			ts.Equal("", poiID)
+			return &NotificationProfile{}, nil
+		})
+
+	ts.env.OnActivity(ts.worker.CheckLocationSpikeActivity, mock.Anything, mock.Anything, mock.Anything).Return(
+		func(ctx context.Context, symptoms []string) ([]schema.Symptom, error) {
+			return []schema.Symptom{}, nil
+		})
+
+	ts.env.ExecuteWorkflow(ts.worker.AccountStateUpdateWorkflow, ts.testAccountNumber)
+
+	ts.env.AssertNumberOfCalls(ts.T(), "CalculateAccountStateActivity", 1)
+	ts.env.AssertNumberOfCalls(ts.T(), "RefreshLocationStateActivity", 1)
+	ts.env.AssertNumberOfCalls(ts.T(), "CheckLocationSpikeActivity", 1)
+	ts.True(ts.env.IsWorkflowCompleted())
+	ts.EqualError(ts.env.GetWorkflowError(), "ContinueAsNew")
+}
+
+// TestAccountStateUpdateWorkflowStateChange validate whether
+// `NotifyLocationStateActivity` is triggered when the state of a location changes
+func (ts *ScoreWorkflowTestSuite) TestAccountStateUpdateWorkflowStateChange() {
+	ts.env.OnActivity(ts.worker.CalculateAccountStateActivity, mock.Anything, mock.Anything).Return(
+		func(ctx context.Context, accountNumber string) (*schema.Metric, error) {
+			ts.Equal(ts.testAccountNumber, accountNumber)
+			return twoSpikeMetric, nil
+		})
+
+	ts.env.OnActivity(ts.worker.RefreshLocationStateActivity, mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(
+		func(ctx context.Context, accountNumber, poiID string, metric schema.Metric) (*NotificationProfile, error) {
+			ts.Equal(ts.testAccountNumber, accountNumber)
+			ts.Equal("", poiID)
+			return &NotificationProfile{
+				StateChangedAccounts: []string{fakeAccount1, fakeAccount2},
+			}, nil
+		})
+
+	ts.env.OnActivity(ts.worker.NotifyLocationStateActivity, mock.Anything, mock.Anything, mock.Anything).Return(
+		func(ctx context.Context, id string, accounts []string) error {
+			ts.Equal("", id)
+			ts.Equal([]string{fakeAccount1, fakeAccount2}, accounts)
+			return nil
+		})
+
+	ts.env.OnActivity(ts.worker.CheckLocationSpikeActivity, mock.Anything, mock.Anything, mock.Anything).Return(
+		func(ctx context.Context, symptoms []string) ([]schema.Symptom, error) {
+			ts.Equal(twoSpikeMetric.Details.Symptoms.LastSpikeList, symptoms)
+			return []schema.Symptom{}, nil
+		})
+
+	ts.env.ExecuteWorkflow(ts.worker.AccountStateUpdateWorkflow, ts.testAccountNumber)
+
+	ts.env.AssertNumberOfCalls(ts.T(), "CalculateAccountStateActivity", 1)
+	ts.env.AssertNumberOfCalls(ts.T(), "RefreshLocationStateActivity", 1)
+	ts.env.AssertNumberOfCalls(ts.T(), "NotifyLocationStateActivity", 1)
+	ts.env.AssertNumberOfCalls(ts.T(), "CheckLocationSpikeActivity", 1)
+
+	ts.True(ts.env.IsWorkflowCompleted())
+	ts.EqualError(ts.env.GetWorkflowError(), "ContinueAsNew")
+}
```

Comment:
Are those activities asynchronized?  That means we can have two activities of the same name run at the same time. If we run numbers of creating several CalculateAccountStateActivity (here, we create 2 times because of the previous Test), are their AssertNumberOfCalls still 1 ?  it seems these test are using the same environment. Make my question simple that is if CalculateAccountStateActivity takes 2 sec to complete, Does these tests still works?

Reply:
Good point. I check the document of the test suite. The whole test cases share a `testing.T`. If it is not set to parallel, the case you mention will no happen. I write test cases in sequential for now. It will work.